pkg/blob/controllerserver.go

```diff
@@ -667,6 +689,107 @@ func (d *Driver) DeleteBlobContainer(ctx context.Context, subsID, resourceGroupN
 	})
 }
 
+// CopyBlobContainer copies a blob container in the same storage account
+func (d *Driver) CopyBlobContainer(ctx context.Context, req *csi.CreateVolumeRequest, accountKey, dstContainerName, storageEndpointSuffix string, secrets map[string]string) error {
+	var volumeID string
+	if req.GetVolumeContentSource() != nil {
+		if req.GetVolumeContentSource().GetVolume() != nil {
+			volumeID = req.GetVolumeContentSource().GetVolume().GetVolumeId()
+		}
+	}
+	resourceGroupName, accountName, srcContainerName, _, subsID, err := GetContainerInfo(volumeID)
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	klog.V(2).Infof("begin to copy blob container %s to %s", srcContainerName, dstContainerName)
+	if srcContainerName == "" {
+		return fmt.Errorf("srcContainerName is empty")
+	}
+	if dstContainerName == "" {
+		return fmt.Errorf("dstContainerName is empty")
+	}
+	if len(secrets) > 0 {
+		accountSasToken, err := getAccountSasToken(secrets, accountName)
+		if err != nil || accountSasToken == "" {
+			klog.V(2).Infof("get account sas token from secrets err: %s, generate sas token for account(%s)", err.Error(), accountName)
+			accountSasToken, err = generateSASToken(accountName, accountKey, storageEndpointSuffix)
+			if err != nil {
+				return err
+			}
+		}
+		srcContainer, getErr := getContainerReference(srcContainerName, secrets, d.cloud.Environment)
+		if getErr != nil {
+			return getErr
+		}
+		exists, err := srcContainer.Exists()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("source container(%s) does not exist", srcContainerName)
+		}
+
+		if err := d.waitForAzCopySuccess(ctx, accountName, resourceGroupName, storageEndpointSuffix, srcContainerName, dstContainerName, accountSasToken, waitForCopyInterval, waitForCopyTimeout); err != nil {
+			return fmt.Errorf("failed to wait for azcopy success: %v", err.Error())
+		}
+	} else {
+		klog.V(2).Infof("generate sas token for account(%s)", accountName)
+		accountSasToken, genErr := generateSASToken(accountName, accountKey, storageEndpointSuffix)
+		if genErr != nil {
+			return genErr
+		}
+		if _, rerr := d.cloud.BlobClient.GetContainer(ctx, subsID, resourceGroupName, accountName, srcContainerName); rerr != nil {
+			return fmt.Errorf("srcContainer(%s) get error: %s", srcContainerName, rerr.Error())
+		}
+
+		if err := d.waitForAzCopySuccess(ctx, accountName, resourceGroupName, storageEndpointSuffix, srcContainerName, dstContainerName, accountSasToken, waitForCopyInterval, waitForCopyTimeout); err != nil {
```

Comment:
this is duplicating code same as L732, and always generate sas token for azcopy, no one will bring sas token for volume cloning now, so we could ignore getAccountSasToken now.

Reply:
remove getAccountSasToken func